File: Algorithmic_Toolbox_C1/W2_Warm_up/fibonacci_last_num.cpp
/*
 ============================================================================
 Name        : fibonacci_last_num.cpp
 Author      : Ivan Arevalo
 Version     :
 Copyright   : Your copyright notice3
 Description : Compute last nth fibonacci number
 ============================================================================
 */

#include <iostream>
#include <cassert>
#include <cstdlib> 
#include <ctime>

#define CHAR_BUFFSIZE (int) (1<<20) // 2^20 = 1048576 bytes around 1 mb


char fib_naive(int n) {
    if (n <= 1)
        return n;
    
    char past = 0;
    char current = 1;
    char temp;

    for(int i = 2; i <= n; ++i){

    	temp = past;
    	past = current;
    	current = (current + temp) % 10;

    }
    return current;
}

char fib_eff(int n) {
	if (n <= 1)
        return n;

    char fib_num[CHAR_BUFFSIZE+1]; fib_num[0] =  0; fib_num[1] = 1;

    for(int i = 2; i <= n; ++i){
    	fib_num[i] = (fib_num[i-1] + fib_num[i-2]) % 10;

    	if(fib_num[i-1] == 0 && fib_num[i] == 1){
    		return fib_num[n % (i-1)];
    	}

    }

    return fib_num[n];
}


void stress_test(int max_n) {

	int counter = 0;
	int n;
	srand(time(0));

	while(1){
		n = rand() % max_n; n++;
		char sol1 = fib_naive(n);
		char sol2 = fib_eff(n);

		assert(sol1 == sol2);
		printf("Test %d successful\n", counter);
		counter++;

	}

}

int main() {
	char debug = 0;

	if(debug){
		stress_test(1000);
	}else{
		int n = 0;
	    std::cin >> n;

	    // std::cout << (int) fib_naive(n) << '\n';
	    std::cout << (int) fib_eff(n) << '\n';
	}

    
    return 0;
}

File: Algorithmic_Toolbox_C1/W2_Warm_up/fibonacci_last_num.py
# python3
# Compute the last digit of the n-th Fibonacci number.

import numpy as np

# Naive implementation, stores whole fibonnaci number, used for stress testing more efficient implementations.
def get_fibonacci_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % 10

# More efficient implementation, stores only the ones place of each fibonacci number.
def get_fibonacci_last_digit_table(n):
    fib_table = [0,1]
    if n<=1:
        return n

    for i in range(2,n+1):
        fib_table.append((fib_table[i-1] + fib_table[i-2]) % 10)

    return fib_table[-1]

# More memory efficient implementation, stores only two variables.
def get_fibonacci_last_digit_mem_eff(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, (previous + current) % 10

    return current

def stress_test(fib_range=20):
    counter = 0
    while(True):
        n = np.random.randint(fib_range+1)

        sol1 = get_fibonacci_last_digit_naive(n)
        sol2 = get_fibonacci_last_digit_table(n)
        assert(sol1 == sol2), f"ERROR: {n}\n SOL1: {sol1}\nSOL2: {sol2}"
        print(f"Test {counter} succesful")
        counter += 1

if __name__ == '__main__':
    debug = False

    if debug:
        stress_test()
    else:
        n = int(input())
        # print(get_fibonacci_remainder_naive(n))
        print(get_fibonacci_last_digit_mem_eff(n))